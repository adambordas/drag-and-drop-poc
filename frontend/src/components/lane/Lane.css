.lane-container {
  flex: 1 1;
  padding: 10px;
}

.lane-container h1 {
  display: block;
  text-align: center;
  font-size: 20px;
}

.lane {
  display: flex;
  flex-direction: column;
  gap: 5px;
  padding: 10px;
  border: 1px solid #DDD;
  min-height: 40px;
}
