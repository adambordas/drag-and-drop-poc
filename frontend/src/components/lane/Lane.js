import './Lane.css';
import Card from '../card/Card'
import { useMutation } from '@apollo/client';
import { MOVE_CARD } from '../../lib/queries';

function Lane({ name, cards }) {
  const [moveCard] = useMutation(MOVE_CARD);

  const allowDrop = e => {
    e.preventDefault();
  };

  const receiveCardDrop = async e => {
    e.preventDefault();

    const cardId = e.dataTransfer.getData('cardId');

    await moveCard({
      variables: {
        id: cardId,
        lane: name
      }
    });
  };
  
  return (
    <div className="lane-container">
      <h1>{name}</h1>
      <div className="lane" onDrop={receiveCardDrop} onDragOver={allowDrop}>
        {
          cards.map(card => <Card key={card.id} id={card.id} description={card.description} />)
        }
      </div>
    </div>
  );
}

export default Lane;
