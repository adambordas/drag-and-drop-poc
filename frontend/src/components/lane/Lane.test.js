import React from 'react';
import { cleanup, fireEvent, render, waitFor } from '@testing-library/react';
import Lane from './Lane';
import { MockedProvider } from '@apollo/client/testing';

const mockMutation = jest.fn();
jest.mock('@apollo/client', () => ({
  useMutation: () => {
    return [mockMutation];
  },
  gql: (str) => {
    return str;
  }
}));

const cardsData = [{
  id: 'test id 1',
  description: 'test description 1',
  lane: 'test lane',
}, {
  id: 'test id 2',
  description: 'test description 2',
  lane: 'test lane',
}, {
  id: 'test id 3',
  description: 'test description 3',
  lane: 'test lane',
}];

const mockPreventDefault = jest.fn();

describe('Lane component', () => {
  beforeEach(() => {
    mockPreventDefault.mockClear();
    mockMutation.mockClear();
  });

  afterEach(() => {
    cleanup();
  });

  it('renders properly', () => {
    const name = 'test name';

    const { container, getByText, queryAllByText } = render(
      <MockedProvider>
        <Lane name={name} cards={cardsData} />
      </MockedProvider>
    );

    const title = getByText(name);
    const cards = queryAllByText(/^test description/);

    expect(title).toBeInTheDocument();
    expect(cards).toHaveLength(cardsData.length);
    expect(container).toMatchSnapshot();
  });

  it('allows drop event by calling preventing default behaviour', () => {
    const { container } = render(
      <MockedProvider>
        <Lane cards={cardsData} />
      </MockedProvider>
    );

    const lane = container.querySelector('.lane');

    const isPrevented = fireEvent.dragOver(lane);

    expect(isPrevented).toEqual(false);
  });

  it('calls card mutation on drop event with proper parameters', async () => {
    const id = 'test card id';
    const name = 'test lane name';
    const dataTransfer = {
      getData: () => id
    };

    const { container } = render(
      <Lane name={name} cards={cardsData} />
    );

    const lane = container.querySelector('.lane');

    fireEvent.drop(lane, { dataTransfer });

    const expectedParameters = {
      variables: {
        id,
        lane: name
      }
    };

    await waitFor(() => expect(mockMutation).toHaveBeenCalledTimes(1));
    await waitFor(() => expect(mockMutation).toHaveBeenCalledWith(expectedParameters));
  });
});
