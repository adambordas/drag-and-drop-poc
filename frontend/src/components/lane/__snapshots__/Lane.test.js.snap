// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lane component renders properly 1`] = `
<div>
  <div
    class="lane-container"
  >
    <h1>
      test name
    </h1>
    <div
      class="lane"
    >
      <div
        class="card"
        draggable="true"
      >
        test description 1
      </div>
      <div
        class="card"
        draggable="true"
      >
        test description 2
      </div>
      <div
        class="card"
        draggable="true"
      >
        test description 3
      </div>
    </div>
  </div>
</div>
`;
